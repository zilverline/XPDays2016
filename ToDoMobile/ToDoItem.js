import React, { Component } from 'react'
import {
  Text,
  TouchableOpacity,
  View
} from 'react-native';

import { styles } from './styles'

export const ToDoItem = (props) => {
  let to_do_item = props.to_do_item
  return <View style={{flexDirection: 'row', justifyContent: 'space-between', alignItems: 'stretch'}}>
    <View style={{flexDirection: 'column', justifyContent: 'space-between', marginRight: 20}}>
      <Text style={styles.item}>{to_do_item.title}</Text>
      <Text style={styles.deadline}>{to_do_item.due_date}</Text>
    </View>
    <TouchableOpacity onPress={() => props.deleteToDoItem(to_do_item)} style={[styles.buttonSmall]}>
      <Text style={[styles.defaultButtonText]}>Delete</Text>
    </TouchableOpacity>
    <TouchableOpacity onPress={() => props.editToDoItem(to_do_item)} style={[styles.buttonSmall]}>
      <Text style={[styles.defaultButtonText]}>Edit</Text>
    </TouchableOpacity>
  </View>
}
