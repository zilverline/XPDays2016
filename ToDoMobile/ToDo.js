/* @flow */

'use strict';

import React, { Component } from 'react';
import {
  ListView,
  Text,
  TextInput,
  TouchableOpacity,
  View
} from 'react-native';

import { styles } from './styles';
import Config from './Config';
import { ToDoItem } from './ToDoItem'

const tenant_id = 1;

export default class ToDo extends Component {

  constructor(props) {
    super(props);
    this.state = {
      dataSource: new ListView.DataSource({
        rowHasChanged: (row1, row2) => row1 !== row2,
      }),
      loaded: false,
      to_do_item: "",
      deadline: "",
      id: ""
    };
  }

  componentDidMount() {
    this.fetchData();
  }

  render() {
    if (!this.state.loaded) {
      return this.renderLoadingView();
    }
    return (
      <View style={styles.container}>
        <Text style={styles.header}>To Do list</Text>
          <View>
            <TextInput
              placeholder={"Item title"}
              style={styles.input}
              onChangeText={(to_do_item) => this.setState({to_do_item})}
              value={this.state.to_do_item}
            />
            <TextInput
              placeholder={"Deadline"}
              style={styles.input}
              onChangeText={(deadline) => this.setState({deadline})}
              value={this.state.deadline}
            />
            {this.renderButton()}
          </View>
          <ListView
            dataSource={this.state.dataSource}
            renderRow={(to_do_item) => <ToDoItem to_do_item={to_do_item} deleteToDoItem={(to_do_item) => this.deleteToDoItem(to_do_item)} editToDoItem={() => this.editToDoItem(to_do_item)} />}
            style={styles.listView}
          />
      </View>
    );
  }

  renderLoadingView() {
    return (
      <View style={styles.container}>
        <Text>
           Loading todo items...
        </Text>
      </View>
    );
  }

  renderButton() {
    return (
      <TouchableOpacity onPress={() => this.saveJob()} style={[styles.button, {alignSelf: 'center'}]}>
        <Text style={[styles.defaultButtonText]}>Save item</Text>
      </TouchableOpacity>
    );
  }

  async fetchData() {
    let response = await fetch(`${Config.current().url}/tenants/${tenant_id}/todo_items`, {
      method: 'get',
      headers: {
        'Content-Type': 'application/json'
      }
    });
    this.updateList(response);
  }

  saveJob() {
    let url;
    if (this.state.id) {
      url = `${Config.current().url}/tenants/${tenant_id}/todo_items/${this.state.id}`;
      this.createUpdateJob(url, 'put');
    }
    else {
      url = `${Config.current().url}/tenants/${tenant_id}/todo_items`;
      this.createUpdateJob(url, 'post');
    }
  }

  async createUpdateJob(url, saveMethod) {
    let response = await fetch(url, {
      method: saveMethod,
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        title: this.state.to_do_item,
        due_date: this.state.deadline
      })
    });
    this.updateList(response);
    this.setState({to_do_item: "", deadline: "", id: ""})
  }

  async deleteToDoItem(to_do_item) {
    let response = await fetch(`${Config.current().url}/tenants/${tenant_id}/todo_items/${to_do_item.id}`, {
      method: 'delete',
      headers: {
        'Content-Type': 'application/json'
      }
    });
    this.updateList(response);
  }

  editToDoItem(to_do_item) {
    this.setState({to_do_item: to_do_item.title, deadline: to_do_item.due_date, id: to_do_item.id});
  }

  async updateList(response) {
    if (response.status === 200) {
      let newList = await response.json();
      this.setState({
        dataSource: this.state.dataSource.cloneWithRows(newList),
        loaded: true,
      });
    }
  }
}
