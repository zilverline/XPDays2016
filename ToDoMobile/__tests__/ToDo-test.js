import 'react-native'
import React from 'react'
import ToDo from '../ToDo'
import { ToDoItem } from '../ToDoItem'

import renderer from 'react-test-renderer'

it('renders shows the loading message', () => {
  const todo = renderer.create(<ToDo />).toJSON();
  expect(todo).toMatchSnapshot();
})

it('renders a to_do_item', () => {
  const todo = renderer.create(
    <ToDoItem
      to_do_item={{id: 1, title: 'Eerste item'}}
      deleteToDoItem={(item) => {}}
      editToDoItem={(item) => {}}
    />
  ).toJSON();
  expect(todo).toMatchSnapshot();
})
