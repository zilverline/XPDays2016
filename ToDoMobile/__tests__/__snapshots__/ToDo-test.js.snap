exports[`test renders a to_do_item 1`] = `
<View
  style={
    Object {
      "alignItems": "stretch",
      "flexDirection": "row",
      "justifyContent": "space-between",
    }
  }>
  <View
    style={
      Object {
        "flexDirection": "column",
        "justifyContent": "space-between",
        "marginRight": 20,
      }
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "paddingTop": 10,
          "textAlign": "left",
        }
      }>
      Eerste item
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "red",
          "textAlign": "left",
        }
      } />
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "black",
        "borderColor": "black",
        "borderRadius": 6,
        "borderWidth": 1,
        "height": 30,
        "margin": 5,
        "opacity": 1,
        "paddingBottom": 10,
        "paddingTop": 5,
        "width": 80,
      }
    }
    testID={undefined}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "gold",
            "fontSize": 14,
            "textAlign": "center",
          },
        ]
      }>
      Delete
    </Text>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "black",
        "borderColor": "black",
        "borderRadius": 6,
        "borderWidth": 1,
        "height": 30,
        "margin": 5,
        "opacity": 1,
        "paddingBottom": 10,
        "paddingTop": 5,
        "width": 80,
      }
    }
    testID={undefined}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "gold",
            "fontSize": 14,
            "textAlign": "center",
          },
        ]
      }>
      Edit
    </Text>
  </View>
</View>
`;

exports[`test renders shows the loading message 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "gold",
      "flex": 1,
      "justifyContent": "center",
      "padding": 50,
    }
  }>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail">
    Loading todo items...
  </Text>
</View>
`;
